sdks/go/pkg/beam/core/runtime/graphx/translate.go

```diff
@@ -1122,7 +1126,12 @@ const defaultEnvId = "go"
 
 func (m *marshaller) addDefaultEnv() string {
 	if _, exists := m.environments[defaultEnvId]; !exists {
-		m.environments[defaultEnvId] = m.opt.Environment
+		env := proto.Clone(m.opt.Environment).(*pipepb.Environment)
+		// Add the pipeline level resource hints here for now.
+		// TODO(https://github.com/apache/beam/issues/23893) move to a better place for
+		// scoped hints in next pass, which affect number of environments set by Go pipelines.
+		env.ResourceHints = m.opt.PipelineResourceHints.Payloads()
```

Comment:
I think `m.opt.PipelineResourceHints` is not guaranteed to be non-nil (e.g. on the direct runner I don't think this will ever get set). It would be good to do a nil check here

Reply:
It isn't guaranteed to be non-nil, but it is implemented to be valid as a zero value. It simply returns an empty map for `Payloads`.

That is, `var Foo Hints` is valid, it's simply an immutable empty value. There is a test to ensure that the zero value Hints is safe for all its methods.

One must use `resource.NewHints` to create a new grouping of hints to ensure these semantics.